cpp/oneapi/dal/backend/primitives/selection/select_by_rows_quick.hpp

```diff
@@ -0,0 +1,240 @@
+
+// Performs k-selection using Quick Select algorithm which is based on row partitioning
+template <typename Float>
+class select_by_rows_quick : public select_by_rows_base<Float> {
+public:
+    select_by_rows_quick() = delete;
+    select_by_rows_quick(sycl::queue& queue, const ndshape<2>& shape) {
+        rnd_seq_ = array<Float>::empty(queue, max_rnd_seq_size_);
+        data_ = ndarray<Float, 2>::empty(queue, shape, sycl::usm::alloc::device);
+        indices_ = ndarray<std::int32_t, 2>::empty(queue, shape, sycl::usm::alloc::device);
+        init(queue, shape[1]);
+    }
+    ~select_by_rows_quick() {
+        last_call_.wait_and_throw();
+    }
+    virtual sycl::event operator()(sycl::queue& queue,
+                                   const ndview<Float, 2>& data,
+                                   std::int64_t k,
+                                   ndview<Float, 2>& selection,
+                                   ndview<std::int32_t, 2>& indices,
+                                   const event_vector& deps) override {
+        return select<true, true>(queue, data, k, selection, indices, deps);
+    }
+    virtual sycl::event operator()(sycl::queue& queue,
+                                   const ndview<Float, 2>& data,
+                                   std::int64_t k,
+                                   ndview<Float, 2>& selection,
+                                   const event_vector& deps) override {
+        ndarray<std::int32_t, 2> dummy;
+        return select<true, false>(queue, data, k, selection, dummy, deps);
+    }
+    virtual sycl::event operator()(sycl::queue& queue,
+                                   const ndview<Float, 2>& data,
+                                   std::int64_t k,
+                                   ndview<std::int32_t, 2>& indices,
+                                   const event_vector& deps) override {
+        ndarray<Float, 2> dummy;
+        return select<false, true>(queue, data, k, dummy, indices, deps);
+    }
+
+private:
+    void init(sycl::queue& queue, std::int64_t rnd_size) {
+        ONEDAL_ASSERT(rnd_size > 1);
+        this->rnd_seq_size_ = rnd_size < select_by_rows_quick::max_rnd_seq_size_
+                                  ? rnd_size
+                                  : select_by_rows_quick::max_rnd_seq_size_;
+        auto engine = daal::algorithms::engines::mcg59::Batch<>::create();
+        auto engine_impl =
+            dynamic_cast<daal::algorithms::engines::internal::BatchBaseImpl*>(&(*engine));
+        ONEDAL_ASSERT(engine_impl != nullptr);
+
+        auto number_array = ndarray<size_t, 1>::empty(queue, this->rnd_seq_size_);
+        daal::internal::RNGs<size_t, daal::sse2> rng;
+        auto number_ptr = number_array.get_mutable_data();
+        rng.uniform((size_t)this->rnd_seq_size_,
+                    number_ptr,
+                    engine_impl->getState(),
+                    0,
+                    (size_t)(this->rnd_seq_size_));
```

Comment:
I suggest to create separate primitive for RNG, but not to use old DAAL API directly